### PR TITLE
SecureBootRecovery: Add Logic to detect 2023 Windows Production CA

### PR DESCRIPTION
**Pull request by Flickdm**
## Description

This change adds the ability for the SecureBootRecovery application to detect the existence of the 2023 Windows Production CA and prevent updating the DB if found.

- [X] Impacts functionality?
  - **Functionality** - Prevents updating the DB if the existence of the 2023 Windows CA is found
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

This application was self signed on a production system using custom Secure Boot keystore. 

This was tested with:

No Secure boot Certificates
Only the 2011 Certificates
All Certificates (2011 and 2023)

## Integration Instructions

N/A

### CODE REVIEW

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (apop5) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: should EFI_UNSUPPORTED get bubbled up through OptionalStatus?

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (apop5) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Should EFI_BUFFER_TOO_SMALL get bubbled up through OptionalStatus?

**Review by apop5 — APPROVED** [review]

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (kenlautner) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Nit: gauranteed

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (kenlautner) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Shouldn't this be `goto Exit`?

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (kenlautner) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Same as above.

**Review by kenlautner — APPROVED** [review]

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: This is actually the expected error, since I'm giving it a NULL buffer with a size of 0. It in return tells me the full size that I can then use to allocate. Are you asking if I should replace EFI_BUFFER_TOO_SMALL with EFI_SUCCESS after this check?

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Why not just use UefiBootServicesTableLib and UefiRuntimeServicesTableLib? They maintain the pointers for you.

I understand dependencies are trying to be reduced but you're already dependent on much more complex logic elsewhere (like the functions backing those pointers) to the point the code redundancy vs complexity tradeoff I think favors just using the MdePkg library instances.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: I suggest moving some of these macros and definitions to a header file.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: I don't know the details of this string's contents, but was there a particular reason for using `UINT8` vs `CHAR8`?

Comment (makubacki) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: It looks like it is expected to be an ASCII string like `"Microsoft Corporation"`? In that case, I'd recommend `CHAR8`.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]:
```suggestion
  OUT EFI_STATUS         *OptionalStatus OPTIONAL

```

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c:70]: I suggest at least an assert to simplify debugging in case of a bad value given to the function.

```suggestion
{
  ASSERT (Message != NULL);
```

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: If I understand the flow correctly, `Db` is no longer needed and the buffer can be freed (on success or through an error path).

```suggestion
Exit:

  if (Db != NULL) {
    mBS->FreePool (Db);
  }

  if (OptionalStatus != NULL) {
    *OptionalStatus = Status;
  }

```

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: In `sizeof (Certificate->Organization)` are you trying to get the size of the struct field (i.e. the pointer) or the size of the string itself?

At a quick glance, it seems that you're trying to get the size of the string here. In that case, `AsciiStrSize()` would be needed to get the size including the NULL character.

This is also where making `Organization` being a `CHAR8` makes it compatible with string functions.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c:70]: If you're trying to completely avoid using `DebugLib`, then just return on NULL.

**Review by makubacki — APPROVED** [review]
Approved assuming noted issues are resolved.

I suggest pinging reviewers when all updates are done so we can take a quick second look before merging.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Since this is going to be consumed by the Boot team and built with windows tooling instead of UEFI ultimately, I'm avoiding the additional libraries

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: This is correct, somehow this got dropped when I was rebasing commits trying to clean up the commits. Thanks!

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: This is getting the size of the string itself. Although in this example I have OrganizationLength and CommonNameLength and don't need to perform the size_of again.

Since the strings lengths are really just a char array which the size of is known at compile time I figured I could avoid pulling in an additional library to get AsciiStrSize(..) and just use str_len

Further, ASN.1 strings (that we're looking for) aren't null terminated and thus we just need the length of the array

Comment (makubacki) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: `OrganizationLength` and `CommonNameLength` are fine to use because they computed the string length using `sizeof` on a byte-sized string literal.

`sizeof (Certificate->Organization)` is the size of the `Organization` member which is `UINT8*` (pointer is 8 bytes on 64-bit machine). `AsciiStrLen()` could be used to not include the null character. But, as you said you can just use the ready-to-go values.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: In that case, I suggest leaving a note saying why this is done since it is against typical edk2 development patterns.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c:29]: I tried breaking up the macros and definitions and I wasn't a fan of breaking it immediately. I might leave this to boot team to figure out how they want to organize it.

Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Moved the certificate defines to a different header file, like you suggested. With a proper comment I like it.

Comment (makubacki) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c]: Yes, it makes it easier to focus on implementation in the C file and quickly find data definitions in the header file.

File: MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c
Comment (makubacki) [inline MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c:214]: Because `DbSize` is an unsigned integer type (`UINTN`), it is always going to be greater than or equal to zero. If you're going to subtract and expect subtraction to potentially go below zero, then this needs to be an `INTN`. If zero is very certain regardless of user input, etc. then this is fine.

Comment (Flickdm) [reply MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c:214]: 0 should be very certain regardless of the user input because it's coming from the GetVariable(..) api. This would mean that the platform is wildly out of spec and I don't think I can handle that.

**Pull request closed without merging**
